Add New Source: MangaIro

getBoolean opened the pull request:
Checklist:
- [x] Updated source's version for individual source changes
- [x] Updated all sources' versions for template changes
- [x] Set appropriate `nsfw` value
- [x] Did not change `id` even if a source's name or language were changed
- [x] Tested the modifications by running it on the simulator or a test device 

Note: Filters do not work with search
Homepage: https://w.mangairo.com/home

Closes https://github.com/Skittyblock/aidoku-community-sources/issues/332

getBoolean commented:
I noticed pagination isn't working, I will work on fixing that.

Skittyblock commented on src/rust/en.mangairo/src/lib.rs:
```diff
@@ -0,0 +1,112 @@
+
+	let total_results = parser::parse_manga_list(html, &mut result);
+
+	if let Some(total_results_value) = total_results {
+		if total_results_value > result.len() as i32 * page {
+			return Ok(MangaPageResult {
+				manga: result,
+				has_more: true,
+			});
+		}
+	}
```
instead of doing all this, I think you could do something like
```rust
let has_more = total_results.map_or(false, |value| value > result.len() as i32 * page);
```

getBoolean replied:
Thanks! I knew there had to be a better way but wasn't sure how to do it in Rust

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,376 @@
+	total_str = total_str.replace(" stories", "");
+	total_str = total_str.chars().filter(|&c| c != ',').collect();
+
+	let total_result = total_str.parse::<i32>();
+	if let Ok(total) = total_result {
+		Some(total)
+	} else {
+		None
+	}
```
```suggestion
	total_str.parse::<i32>().ok()
```

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,376 @@
+		let url = obj.attr("href").read();
+		let id = parse_incoming_url_chapter_id(url.clone());
+
+		if let Some(id_value) = id {
```
You don't need to rename things when you unwrap them
```suggestion
		if let Some(id) = id {
```

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,376 @@
+		if let Some(id_value) = id {
+			let split = id_value.split('-');
+			let vec = split.collect::<Vec<&str>>();
+			let chap_num = vec[vec.len() - 1].parse().unwrap();
```
just call this `chapter`, and instead of unwrapping you should probably just use `?` to return or fallback to `-1` for no chapter number

getBoolean replied:
I've updated it to use `unwrap_or(-1.0)`

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,376 @@
+pub fn get_filtered_url(filters: Vec<Filter>, page: i32, url: &mut String) {
+	let mut is_searching = false;
+	let mut search_string = String::new();
+	url.push_str("https://w.mangairo.com");
```
probably use `BASE_URL` here instead?

Skittyblock reviewed (changes requested):
- icon looks pretty low quality, probably find a better one or use https://github.com/tachiyomiorg/tachiyomi-extensions/blob/651e5cf50754700419c190368a81762e71fcb077/multisrc/overrides/mangabox/mangairo/res/web_hi_res_512.png
- jsons should have newlines at the end of the file
- run `cargo clippy` and fix the warnings and then `cargo fmt` to format

getBoolean commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,376 @@
+	string
+}
+
+pub fn urlencode(string: String) -> String {
```
This URL encode is copied from the website's js code and converted to rust to mimic the same functionality. I do not believe it supports traditional url encoding.

getBoolean replied:
For reference, URI escaping the space yields no results but it does work with underscores.

* https://w.mangairo.com/list/search/solo%20leveling
* https://w.mangairo.com/list/search/solo_leveling

Skittyblock replied:
oh okay, fair enough

getBoolean commented on src/rust/en.mangairo/src/lib.rs:
```diff
@@ -0,0 +1,112 @@
+	let mut result: Vec<Manga> = Vec::new();
+
+	let mut url = String::new();
+	parser::get_filtered_url(filters, page, &mut url);
```
No particular reason, much of lib.rs was copied from the mangapill extension.

I like that much better, I will do that.

getBoolean commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,376 @@
+	let genre_filter: Option<Filter> = filters
+		.iter()
+		.find(|&x| x.kind == FilterType::Select && x.name == "Genre")
+		.cloned();
```
I initially did this so I could ensure the URL was constructed in the correct order. I've refactored this function to use match for the filters and pull out the info needed. I've also removed the mutable strings in this function except for `search_string` and refactored the filter matches to support that.

```rust
	let mut title_filter = String::new();
	let mut author_filter = String::new();
	let mut sort_filter = -1;
	let mut genre_filter = -1;
	let mut status_filter = -1;
	for filter in filters {
		match filter.kind {
			FilterType::Title => {
				if let Ok(filter) = filter.value.as_string() {
					title_filter = urlencode(filter.read().to_lowercase());
					is_searching = true;
				}
			}
			...
```

getBoolean replied:
I am not sure if the mutable variables approach is the best way to do this, but I'm not aware of a better option

getBoolean commented:
Thanks for the thorough review! I especially appreciate the small tips you gave that might otherwise have been overlooked by others.

I've addressed all the feedback comments, but I'd like you to take a look at the fixes to make sure I addressed them properly.

getBoolean commented:
I scaled down that icon you linked to 128x128 from 512x512. The app currently only supports 128x128, correct?

Skittyblock commented on src/rust/en.mangairo/res/settings.json:
no need to have this empty file, you can just remove it

Skittyblock replied:
still needs to be removed

Skittyblock commented on src/rust/en.mangairo/res/source.json:
```diff
@@ -0,0 +1,25 @@
+		"lang": "en",
+		"name": "MangaIro",
+		"version": 1,
+		"url": "https://w.mangairo.com/home",
```
In order for deep links to work and such this should be the base url, even if it isn't the home listing page
```suggestion
		"url": "https://w.mangairo.com",
```

Skittyblock commented on src/rust/en.mangairo/src/lib.rs:
```diff
@@ -0,0 +1,84 @@
+	let parsed_manga_id = parser::parse_incoming_url_manga_id(url.clone());
+	let parsed_chapter_id = parser::parse_incoming_url_chapter_id(url);
+
+	if parsed_manga_id.is_none() {
```
typically best to unwrap by doing something like, rather than calling `unwrap()`
```rust
if let Some(parsed_manga_id) = parsed_manga_id {
    Ok(DeepLink { ... })
} else {
    Err(...)
}
```

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,354 @@
+	}
+
+	// Example: 'Total: 38,202 stories'
+	let mut total_str = html.select(".quantitychapter").text().read();
+	total_str = total_str.replace("Total: ", "");
+	total_str = total_str.replace(" stories", "");
+	total_str = total_str.chars().filter(|&c| c != ',').collect();
```
this doesn't need to be mutable, you could do all the calls on one variable (including the i32 parse)

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,354 @@
+		.read()
+		.to_lowercase();
+
+	let url = format!("{}", &id);
```
ids should **not** be full urls. the best way to handle this is to string the url out so you just have the `/relative/path`, and then you can prepend the `BASE_URL` to get the full url. This makes it easier to fix if the site ever changes domains.

getBoolean replied:
I ran into a problem with that because the latest manga use URL `w.mangairo.com`, but the older manga use `chap.mangairo.com`. I don't remember exactly what was the problem I ran into, I'll revisit this again.

Skittyblock replied:
is it possible to only use `chap.mangairo.com` then?

getBoolean replied:
Should be... I'll need to do some refactoring though.

getBoolean replied:
Unfortunately, it doesn't work. `https://chap.mangairo.com/story-de293390/chapter-1` will redirect to `https://w.mangairo.com/story-de293390`, breaking method `get_page_list`

Skittyblock replied:
alright then it's probably fine to just leave it with the full urls

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,354 @@
+		let id = parse_incoming_url_chapter_id(url.clone());
+
+		if let Some(id) = id {
+			let split = id.split('-');
+			let vec = split.collect::<Vec<&str>>();
+			let chapter: f32 = vec[vec.len() - 1].parse().unwrap_or(-1.0);
```
I think you can do something like this instead
```suggestion
			let chapter = id
				.rsplit_once('-')
				.map(|v| v.1.parse::<f32>().ok())
				.flatten()
				.unwrap_or(-1.0);
```

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,354 @@
+	str
+}
+
+fn replace_consecutive_underscores(input: String) -> String {
```
I'm a little confused. Isn't this the same as `.replace("__", "_")`?

getBoolean replied:
That would collapse three underscores `___` to `_`? I was thinking it would not, but now that I think about it again it makes sense it would.

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,354 @@
+
+	let url = format!("{}", &id);
+
+	let mut authors: Vec<String> = Vec::new();
+	html.select(".story_info_right li:nth-child(3) a")
+		.array()
+		.for_each(|tag| {
+			authors.push(String::from(
+				tag.as_node().expect("node array").text().read().trim(),
+			))
+		});
+	let author = authors.join(", ");
```
I think you could do this
```suggestion
	let author = html
		.select(".story_info_right li:nth-child(3) a")
		.array()
		.map(|tag| {
			String::from(
				tag.as_node().expect("node array").text().read().trim(),
			)
		})
		.join(", ");
```

Skittyblock commented on src/rust/en.mangairo/src/parser.rs:
```diff
@@ -0,0 +1,354 @@
+		});
+	let author = authors.join(", ");
+
+	let mut categories: Vec<String> = Vec::new();
+	html.select(".story_info_right .a-h")
+		.array()
+		.for_each(|tag| categories.push(tag.as_node().expect("node array").text().read()));
```
again, just map these instead of pushing to a mutable vector

Skittyblock commented:
> I scaled down that icon you linked to 128x128 from 512x512. The app currently only supports 128x128, correct?

the app supports anything, but we use 128x128 conventionally since that's the largest the source icon would ever need to be and it's best to be consistent.

Skittyblock commented on src/rust/en.mangairo/res/source.json:
```diff
@@ -0,0 +1,25 @@
+		"lang": "en",
+		"name": "MangaIro",
+		"version": 1,
+		"url": "https://w.mangairo.com",
```
In order to get deeplinks to work
```suggestion
		"urls": [
			"https://w.mangairo.com",
			"https://chap.mangairo.com"
		],
```
